app/src/main/java/com/celzero/bravedns/service/ConnectionMonitor.kt

```diff
@@ -0,0 +1,240 @@
+        override fun handleMessage(msg: Message) {
+            // isForceUpdate - true if onUserPreferenceChanged is changes, the messages should be
+            // processed forcefully regardless of the current and new networks.
+            val isForceUpdate : Boolean = msg.obj as Boolean
+            when(msg.what) {
+                MSG_ADD_ACTIVE_NETWORK -> processActiveNetwork(isForceUpdate)
+                MSG_ADD_ALL_NETWORKS -> processAllNetworks(isForceUpdate)
+            }
+        }
+
+        /**
+         * Handles the active network.
+         * Get active network from connectivity-manager
+         * Set the vpn underlying network when the current active network is different from
+         * already assigned active network or if there is difference in the available network set
+         */
+        private fun processActiveNetwork(isForceUpdate : Boolean){
+            val newActiveNetwork: Network? = connectionMonitor.connectivityManager.activeNetwork
+            val isNewNetwork = connectionMonitor.currentActiveNetwork?.networkHandle != newActiveNetwork?.networkHandle
+            if (DEBUG) Log.d(LOG_TAG, "ConnectionMonitor - ${connectionMonitor.connectivityManager.getNetworkInfo(newActiveNetwork)?.typeName.toString()}, $isNewNetwork")
+            connectionMonitor.currentActiveNetwork = newActiveNetwork
+            connectionMonitor.currentNetworks.clear()
+            if(!isNewNetwork && !isForceUpdate){
+                return
+            }
+            if (connectionMonitor.currentActiveNetwork != null) {
+                if (DEBUG) Log.d(LOG_TAG, "ConnectionMonitor - process message onNetworkConnected- $newActiveNetwork")
+                // Setting the setUnderLyingNetwork to null which by default will set the
+                // active network.
+                connectionMonitor.networkListener.onNetworkConnected(null)
+            } else {
+                if (DEBUG) Log.d(LOG_TAG, "ConnectionMonitor - process message onNetworkDisconnected")
+                connectionMonitor.networkListener.onNetworkDisconnected()
+            }
+            Log.i(LOG_TAG, "ConnectionMonitor - currentActiveNetwork- ${connectionMonitor.connectivityManager.getNetworkInfo(newActiveNetwork)?.typeName.toString()}")
+        }
+
+        /**
+         * Adds all the available network to the underlying network.
+         */
+        private fun processAllNetworks(isForceUpdate : Boolean) {
+            val newActiveNetwork = connectionMonitor.connectivityManager.activeNetwork
+            var isNewNetwork = connectionMonitor.currentActiveNetwork?.networkHandle != newActiveNetwork?.networkHandle
+            val newNetworks = createNetworksSet(newActiveNetwork)
+            isNewNetwork = isNewNetwork || hasDifference(connectionMonitor.currentNetworks, newNetworks)
+            connectionMonitor.currentActiveNetwork  = newActiveNetwork
+            connectionMonitor.currentNetworks = newNetworks
```

Comment:
see above, we may not really need to maintain a `currentActiveNetwork` variable at all.

Reply:
Removed `currentActiveNetwork `